src/disk_io_thread.cpp

```diff
@@ -2319,22 +2326,25 @@ namespace libtorrent
 		// to keep the cache footprint low, try to evict a volatile piece
 		m_disk_cache.try_evict_one_volatile();
 
+		// save a local copy of offset to avoid concurrent access
+		int offset = ph->offset;
+
 		l.unlock();
 
 		int ret = 0;
 		int next_locked_block = 0;
-		for (int i = ph->offset / block_size; i < blocks_in_piece; ++i)
+		for (int i = offset / block_size; i < blocks_in_piece; ++i)
 		{
 			file::iovec_t iov;
-			iov.iov_len = (std::min)(block_size, piece_size - ph->offset);
+			iov.iov_len = (std::min)(block_size, piece_size - offset);
 
 			if (next_locked_block < num_locked_blocks
 				&& locked_blocks[next_locked_block] == i)
 			{
 				++next_locked_block;
 				TORRENT_PIECE_ASSERT(pe->blocks[i].buf, pe);
-				TORRENT_PIECE_ASSERT(ph->offset == i * block_size, pe);
-				ph->offset += int(iov.iov_len);
+				TORRENT_PIECE_ASSERT(offset == i * block_size, pe);
+				offset += int(iov.iov_len);
 				ph->h.update({pe->blocks[i].buf, iov.iov_len});
```

Comment:
there better only be a single thread calling ``update()`` here. that's guaranteed, right?